spec.html

```diff
@@ -11982,12 +11982,12 @@ <h1>InitializeHostDefinedRealm ( ): either a normal completion containing ~unuse
   <emu-clause id="sec-agents">
     <h1>Agents</h1>
 
-    <p>An <dfn id="agent" variants="agents">agent</dfn> comprises a set of ECMAScript execution contexts, an execution context stack, a running execution context, an <dfn id="agent-record" variants="Agent Records">Agent Record</dfn>, and an <dfn id="executing-thread" variants="executing threads">executing thread</dfn>. Except for the executing thread, the constituents of an agent belong exclusively to that agent.</p>
-    <p>An agent's executing thread executes a job on the agent's execution contexts independently of other agents, except that an executing thread may be used as the executing thread by multiple agents, provided none of the agents sharing the thread have an Agent Record whose [[CanBlock]] field is *true*.</p>
+    <p>An <dfn id="agent" variants="agents">agent</dfn> comprises a set of ECMAScript execution contexts, an execution context stack, a running execution context, an <dfn id="agent-record" variants="Agent Records">Agent Record</dfn>, and an <dfn id="executing-thread" variants="executing threads">executing thread</dfn>. Except for the executing thread, Shared Data Blocks, and specification values related to Shared Data Blocks, the constituents of an agent belong exclusively to that agent.</p>
+    <p>An agent's executing thread executes code on the agent's execution contexts independently of other agents, except that an executing thread may be used as the executing thread by multiple agents, provided none of the agents sharing the thread have an Agent Record whose [[CanBlock]] field is *true*.</p>
```

Comment:
Could you explain your thinking behind this change and the one below?

I tried writing a comment, but I'm having trouble without knowing your intent.

Reply:
The intention is to make this section more generic than about jobs. The history wrt jobs IIRC was that we had a very generic job queue scheduler system that no embedder actually used. This was dead code that no host used, so it was very misleading, and we removed it. So when agents talk about "jobs" I'm pretty sure they're talking about this old notion of a job that has been ripped out.

Since then we've redefined a "Job" to be something more specific: an Abstract Closure that's given to the host to schedule and run. So it seems odd to me to say that "surrounding agent" is only well-defined when an agent is executing a Job. The initial script evaluation, for instance, is not part of a Job in the Abstract Closure sense.

So my intention here was to make the notion of agent executing JS code to be more broad than just the restricted notion of Job. I think "executes code" suffices to broad enough. I had originally thought something like "evaluating ECMAScript code", but I want the "surrounding agent" notion to be well-defined in the waitAsync timer stuff, which is arguably not executing ECMAScript code. Thoughts?

Reply:
Thanks.

> I think "executes code" suffices to broad enough. I had originally thought something like "evaluating ECMAScript code", but I want the "surrounding agent" notion to be well-defined in the waitAsync timer stuff, which is arguably not executing ECMAScript code. Thoughts?

In the spec, the word "code" (outside of "code unit" and "code point") almost always means ECMAScript source code anyway, so "executes code" doesn't have the meaning you want. I think "executes algorithmic steps" would be better. (You could maybe drop the "algorithmic" for subsequent uses.)

```suggestion
    <p>An agent's executing thread executes algorithmic steps on the agent's execution contexts independently of other agents, except that an executing thread may be used as the executing thread by multiple agents, provided none of the agents sharing the thread have an Agent Record whose [[CanBlock]] field is *true*.</p>
```

or maybe "executes algorithms".

----

> The intention is to make this section more generic than about jobs.

I'm okay with the idea that an agent can execute a slightly broader class of thing than just jobs, but I'd like it to be clear that a job is one of the things it can execute. Currently, the only explicit connection between 'agent' and 'job' is in this para and the next, and this PR is dissolving that connection. Maybe somewhere near the definition of job, it could say that a job is executed by an agent. (See also issue #2642.)

> The history wrt jobs IIRC was that we had a very generic job queue scheduler system that no embedder actually used. This was dead code that no host used, so it was very misleading, and we removed it.

Yup, that was PR #1597.

> So when agents talk about "jobs" I'm pretty sure they're talking about this old notion of a job that has been ripped out.

Well, except #1597 also changed these paragraphs, so presumably they were modified to talk about the new notion of a job.

> Since then we've redefined a "Job" to be something more specific: an Abstract Closure that's given to the host to schedule and run.

That was also done in #1597.

> So it seems odd to me to say that "surrounding agent" is only well-defined when an agent is executing a Job. The initial script evaluation, for instance, is not part of a Job in the Abstract Closure sense.

I agree it's odd. (#1597 dropped ScriptEvaluationJob without saying that an agent could execute non-jobs like ScriptEvaluation.)

But if an agent can execute non-jobs, are there any restrictions on when/how it executes them? (E.g., can an agent start executing a ScriptEvaluation while another is in progress?)

Reply:
I've added a bit to the Job section about being executed by agents.